Mutex around writer access in recorder

Stapelzeiger opened the pull request:
The storage plugins are not designed to be thread safe. The recorder must therefore use a mutex when accessing the writer from different contexts. This happens for topic creation during discovery as well as in message callbacks from subscribed topics.

This fixes #486 and #480.

dawonn-haval commented:
Just wanted to leave a note that this patch seems to have made an improvement on my system. Before this patch I was only able to record sensor data bag files for a 5~10 seconds before it would hang and stop recording, but now rosbag constantly runs 1~2 minutes before hanging! 

FWIW: Foxy, Ubuntu 20.04, 5 lidar + 1 camera.

Stapelzeiger commented:
@Karsten1987 I added locking in the `rosbag2_cpp` writer. I've reverted the changes in recorder, since they are redundant with the `rosbag2_cpp` ones so this should solve your second point regarding the `create_subscription` in the critical zone.

Do we need a similar thread-safe interface for the reader class? I have trouble imagining a use-case where you would read the from multiple threads with the current API (in particular the `has_next()`, `read_next()` wouldn't work nicely).

Karsten1987 commented on rosbag2_cpp/include/rosbag2_cpp/writer.hpp:
```diff
@@ -19,6 +19,7 @@
 #include <string>
 #include <unordered_map>
 #include <vector>
+#include <mutex>
```
alpha sort

jarledo commented:
Any updates on this? Our system suffers from the bug described in #486, making it almost impossible to record our topics. Would be great to have this merged.

Karsten1987 commented:
Thanks for the reminder, kicking off CI on this:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12823)](http://ci.ros2.org/job/ci_linux/12823/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7780)](http://ci.ros2.org/job/ci_linux-aarch64/7780/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10537)](http://ci.ros2.org/job/ci_osx/10537/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12768)](http://ci.ros2.org/job/ci_windows/12768/)

Karsten1987 commented:
@Stapelzeiger thanks for the contribution

Karsten1987 commented:
@adamdbrw fyi

almaslov commented:
Hello! It would be great to backport this fix to Foxy. Can I do it myself according to [the Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#guidelines-for-backporting-prs) ?

emersonknapp commented:
Yes, by all means. If the cherry-pick works without issue then it is an easy backport.
